Change sold2 detector config to dataclass

lappemic opened the pull request:
#### Changes
So far, I've introduced dataclasses for configs in `kornia.feature.sold2.sold2_detector`. Here's a quick look:

- `HeatMapRefineCfg`, `JunctionRefineCfg`, `LineDetectorCfg`, and `BackboneCfg` are now dataclasses
- Nested configs like `heatmap_refine_cfg` in `LineDetectorCfg` use `default_factory` for safe defaults
- Introduced dict_to_dataclass() to convert from ones dict to the corresponding dataclass

Is this the right direction? Can I go on with this approach? 


[Update configs to use dataclasses](https://github.com/kornia/kornia/issues/2093#top) -- related to #2093

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 📝 This change requires a documentation update


#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

johnnv1 reviewed (commented):
Yes this is the direction, at the end of this PR (or later) we can move these useful functions like `dict_to_dataclass` to somewhere in `kornia.utils`

lappemic commented:
Hey @johnnv1, I've made the following two updates and would love your input:
- Added a `dataclass_to_dict` function for converting the new dataclass configs into dictionaries when necessary
- Tweaked `SOLD2_detector` so it works with these dataclass configs (was thinking about updating 'SOLD2Net` to directly accept dataclasses but simplicitywise for this PR skipped it.)

Is the approached implementation what you are looking for? Your thoughts would be super helpful! Once I hear back, I'll move on to adjusting `LineSegmentDetectionModule` too. Thanks!

johnnv1 commented:
> Is the approached implementation what you are looking for?


Yes it's

lappemic commented:
Hey @johnnv1 i implemented the rest as i thought it might fit best. The following changes are in the last couple commits:

- **Removed `default_detector_cfg` dictionary**: Not used anymore as intended
- **Extended `LineDetectorCfg`**: Now includes all configurations used in `LineSegmentDetectionModule`
- **Updated initialization and documentation**: 
  - Modified `LineSegmentDetectionModule` to directly use the `LineDetectorCfg` dataclass
  - Updated docstrings in `LineSegmentDetectionModule` and `SOLD2_detector` to reflect these changes
- **Removed `dict_to_dataclass` utility**: This is actually not needed
- **Updated method call**: Changed `LineSegmentDetectionModule` instantiation to pass the dataclass directly

I think the changes should be ok, but I’m looking forward to your feedback and am happy to adjust where necessary.

johnnv1 commented on kornia/feature/sold2/sold2_detector.py:
```diff
@@ -61,25 +100,24 @@ class SOLD2_detector(Module):
     """
 
-    def __init__(self, pretrained: bool = True, config: Optional[Dict[str, Any]] = None) -> None:
+    def __init__(self, pretrained: bool = True, config: Optional[DetectorCfg] = None) -> None:
```
I think it's ok if we drop support for the dict as config on the typing already, but over the functionality we should keep support with a warning, something like

`import warnings`
```suggestion
    def __init__(self, pretrained: bool = True, config: Optional[DetectorCfg] = None) -> None:
        if isinstance(cofig, dict):
            warnings.warn(
                    f"Usage of config as a plain dictionary is deprecated in favor of `kornia.feature.sold2.sold2_detector.DetectorCfg`. The support of plain dictionaries as config will be removed in kornia v0.8.0 (December 2024).",
                    category=DeprecationWarning,
                    stacklevel=2,
                )
            config = dict_to_dataclass(**config, DetectorCfg)
```

johnnv1 reviewed (changes requested):
Can you fix the typing issue? Also, the tests of Sold2 now are falling, you can run it with `pytest tests/feature/test_sold2.py --runslow`

lappemic commented:
Thanks @johnnv1. I fixed the typo and the pytests pass local. There failing `typing` issue i am not sure how to resolve. I am not really experienced with `typing`, if you could give me some inputs on this would be great. 

The rest seems good to go to me. WDYT?

johnnv1 commented on kornia/feature/sold2/sold2_detector.py:
```diff
@@ -14,31 +16,82 @@
+    else:
+        return obj
+
+
+def dict_to_dataclass(dict_obj: Dict[str, Any], dataclass_type: Type[Any]) -> Any:
```
```suggestion
from typing import TypeVar

T = TypeVar('T')

def dict_to_dataclass(dict_obj: Dict[str, Any], dataclass_type: T) -> T:
```

johnnv1 commented on kornia/feature/sold2/sold2_detector.py:
```diff
@@ -14,31 +16,82 @@
+    line_detector_cfg: LineDetectorCfg = field(default_factory=LineDetectorCfg)
+
+
+def dataclass_to_dict(obj: Any) -> Union[Dict[str, Any], List[Any], Tuple[Any, ...]]:
```
Could you move this one, and the `dict_to_dataclass` to the file "kornia/utils/helpers.py"?

lappemic replied:
Moved the `dataclass_to_dict` and the `dict_to_dataclass` functions to `kornia/utils/helpers.py`.

Let me know if there is another thing i can adjust or improve.

johnnv1 commented on kornia/utils/helpers.py:
```diff
@@ -318,3 +319,34 @@ def is_autocast_enabled(both: bool = True) -> bool:
+        return obj
+
+
+T = TypeVar("T", bound=Type[Any])
+
+
+def dict_to_dataclass(dict_obj: Dict[str, Any], dataclass_type: T) -> Any:
+    """Recursively convert dictionaries to dataclass instances."""
+    if not isinstance(dict_obj, dict):
+        raise TypeError("Input conf must be dict")
+    if not is_dataclass(dataclass_type):
+        raise TypeError("dataclass_type must be a dataclass")
+    field_types = {f.name: f.type for f in fields(dataclass_type)}
+    constructor_args = {}
+    for key, value in dict_obj.items():
+        if key in field_types and is_dataclass(field_types[key]):
+            constructor_args[key] = dict_to_dataclass(value, field_types[key])
+        else:
+            constructor_args[key] = value
+    return dataclass_type(**constructor_args)
```
```suggestion
T = TypeVar("T")


def dict_to_dataclass(dict_obj: Dict[str, Any], dataclass_type: Type[T]) -> T:
    """Recursively convert dictionaries to dataclass instances."""
    if not isinstance(dict_obj, dict):
        raise TypeError("Input conf must be dict")
    if not is_dataclass(dataclass_type):
        raise TypeError("dataclass_type must be a dataclass")
    field_types = {f.name: f.type for f in fields(dataclass_type)}
    constructor_args = {}
    for key, value in dict_obj.items():
        if key in field_types and is_dataclass(field_types[key]):
            constructor_args[key] = dict_to_dataclass(value, field_types[key])
        else:
            constructor_args[key] = value
    # TODO: remove type ignore when https://github.com/python/mypy/issues/14941 be andressed
    return dataclass_type(**constructor_args)  # type: ignore[return-value]

```

johnnv1 reviewed (approved):
Thank you for this patch, can you address https://github.com/kornia/kornia/issues/2093#issuecomment-2059102827 in a follow-up PR?

lappemic commented:
As soon as this PR is merged, i will rework the dataclasses in a `structure.py` file in a separate PR.

lappemic commented:
Hey @johnnv1, do i need to make some changes so we can merge this PR?

johnnv1 commented:
> Hey @johnnv1, do i need to make some changes so we can merge this PR?
only have green lights on CI, not sure if the CI docs failed randomly or not, I rerun it now

lappemic commented:
Do we need to do something about the failed docs @johnnv1?
